### PR TITLE
Miscellaneous fixes for acceptance test failures

### PR DESCRIPTION
**Pull request by jglick**
Partly follows up #301, plus some other fixes inspired by looking at a thread dump from a `linearFlow` failure. Per-commit messages have detailed justifications.

Now subsumes #310.

@reviewbybees

### CODE REVIEW

File: cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 1]
```diff
@@ -854,12 +856,16 @@ void notifyListeners(List<FlowNode> nodes, boolean synchronous) {
     }
 
     @Restricted(DoNotUse.class)
-    @Terminator public static void suspendAll() throws InterruptedException, ExecutionException {
+    @Terminator public static void suspendAll() throws InterruptedException, ExecutionException, TimeoutException {
```
Comment (oleg-nenashev) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: It's binary incompatible

**Comment by rsandell** [comment]
:bee:

File: cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 1 above)
Comment (jglick) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: This is explicitly not offered as an API.

Comment (oleg-nenashev) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Uhm, right

**Comment by oleg-nenashev** [comment]
:bee:

**Comment by recena** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
Pretty much workaround magic, but :bee:

**Comment by christ66** [comment]
:bee:

File: cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 2]
```diff
@@ -854,12 +856,16 @@ void notifyListeners(List<FlowNode> nodes, boolean synchronous) {
...
         LOGGER.fine("starting to suspend all executions");
         for (FlowExecution execution : FlowExecutionList.get()) {
             if (execution instanceof CpsFlowExecution) {
                 LOGGER.log(Level.FINE, "waiting to suspend {0}", execution);
-                ((CpsFlowExecution) execution).waitForSuspension();
+                CpsFlowExecution exec = (CpsFlowExecution) execution;
+                // Like waitForSuspension but with a timeout:
+                if (exec.programPromise != null) {
+                    exec.programPromise.get(1, TimeUnit.MINUTES).scheduleRun().get(1, TimeUnit.MINUTES);
+                }
```
Comment (jglick) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Should have fixed [JENKINS-28221](https://issues.jenkins-ci.org/browse/JENKINS-28221).

**Pull request merged**